### PR TITLE
Improve Directory tree performance of New Media Gallery

### LINKED ISSUE #30596: Magento 2.4.1 Media Gallery....not ready for prime time. Fails when you have large amounts of data (images)
### Preconditions (*)

The new media gallery introduced in Magento 2.4.1 certainly was not vetted using a large robust shop. Maybe it was designed with 100-200 products like the using Magento sample data but I can tell you from my experience trying to use this, it does NOT work when you have a large amount images. Since we use compression software to have images ready for desktop, table, phone and WEBP use.  We have generated over a half a million images for our shop. What happens is that the images will load but the directory tree fails to generate due to sessions being destroyed and memory exhaustion. Even setting php.ini to basically unlimited use of ram and setting our instance to 192 GB physical ram, the media gallery still fails to load the directory tree. This is very disappointing! What I don't understand either is why would the media galley be set with  /media as the storage root? This is the main problem as 95% of the images in the media gallery are product images. Most shops don't use product images to create banners or make CMS blocks or pages. The correct root for media gallery in my opinion should have been the /wysiwyg directory. That is what is used when dealing with media and the CMS block and pages creation. If this was set as root then these issues with session destruction and memory failures would never happen. Or, the media gallery should exempt the catalog/product images like it is in the old media gallery. If you go to those catalog/product folders you will see there are no images. That is another correct way this gallery should have been setup, in my humble opinion.

### How to produce error (*)
1. Install M 2.4.1
2. Move your media content from your current or old website that has 500,000 or more images
3. Go to content/media galley
4. Media gallery will open with images but directory tree is missing.  You can't find folders or see any folders or even see new ones if you make any. Note the folders are created but because the tree is not there, you cannot see the folders. Also it is random whether the search, sort or other fields show up. Sometimes they do and sometimes they dont.

### Steps to reproduce (*)

1. Install M 2.4.1
2. Move your media content from your current or old website that has 500,000 or more images
3. Go to content/media galley
4. Open Media gallery 

### Expected result (*)

1. I expect that the directory tree, search, sort and full media gallery opens, not a crippled version of the media gallery that does not even work.

see various images. It should look like this:

image

### Actual result (*)

1. Many errors in php log:

[19-Oct-2020 04:26:51 UTC] PHP Fatal error: Allowed memory size of 792723456 bytes exhausted (tried to allocate 8192 bytes) in /home/cloudpanel/htdocs/www.performance-pcs.com/vendor/magento/framework/Filesystem/Driver/File.php on line 962
[19-Oct-2020 04:26:51 UTC] PHP Fatal error: Allowed memory size of 792723456 bytes exhausted (tried to allocate 135168 bytes) in /home/cloudpanel/htdocs/www.performance-pcs.com/vendor/magento/framework/Session/SessionManager.php on line 151
[19-Oct-2020 04:26:51 UTC] PHP Fatal error: Allowed memory size of 792723456 bytes exhausted (tried to allocate 65536 bytes) in /home/cloudpanel/htdocs/www.performance-pcs.com/vendor/colinmollenhour/php-redis-session-abstract/src/Cm/RedisSession/Handler.php on line 793
[19-Oct-2020 04:26:51 UTC] PHP Fatal error: Uncaught RuntimeException: Unknown: Cannot call session save handler in a recursive manner in [no active file]:0[19-Oct-2020 08:29:18 UTC] PHP Warning: DOMXPath::query(): Invalid expression in /home/cloudpanel/htdocs/www.performance-pcs.com/vendor/pelago/emogrifier/src/Emogrifier.php on line 494
[19-Oct-2020 08:29:18 UTC] PHP Warning: DOMXPath::query(): Invalid expression in /home/cloudpanel/htdocs/www.performance-pcs.com/vendor/pelago/emogrifier/src/Emogrifier.php on line 494
[19-Oct-2020 08:29:18 UTC] PHP Warning: Invalid argument supplied for foreach() in /home/cloudpanel/htdocs/www.performance-pcs.com/vendor/pelago/emogrifier/src/Emogrifier.php on line 503
[19-Oct-2020 09:06:41 UTC] PHP Fatal error: Allowed memory size of 792723456 bytes exhausted (tried to allocate 8192 bytes) in /home/cloudpanel/htdocs/www.performance-pcs.com/vendor/magento/framework/Filesystem/Driver/File.php on line 962
[19-Oct-2020 09:06:41 UTC] PHP Fatal error: Allowed memory size of 792723456 bytes exhausted (tried to allocate 40960 bytes) in /home/cloudpanel/htdocs/www.performance-pcs.com/vendor/magento/framework/Session/SessionManager.php on line 151
[19-Oct-2020 09:06:41 UTC] PHP Fatal error: Allowed memory size of 792723456 bytes exhausted (tried to allocate 65536 bytes) in /home/cloudpanel/htdocs/www.performance-pcs.com/vendor/colinmollenhour/php-redis-session-abstract/src/Cm/RedisSession/Handler.php on line 793
[19-Oct-2020 09:06:41 UTC] PHP Fatal error: Uncaught RuntimeException: Unknown: Cannot call session save handler in a recursive manner in [no active file]:0[19-Oct-2020 09:54:47 UTC] PHP Fatal error: Uncaught RuntimeException: ObjectManager isn't initialized in /home/cloudpanel/htdocs/www.performance-pcs.com/vendor/magento/framework/App/ObjectManager.php:36
Stack trace:
#0 /home/cloudpanel/htdocs/www.performance-pcs.com/vendor/magento/framework/Filesystem/Directory/Read.php(78): Magento\Framework\App\ObjectManager::getInstance()
#1 /home/cloudpanel/htdocs/www.performance-pcs.com/vendor/magento/framework/Filesystem/Directory/Read.php(212): Magento\Framework\Filesystem\Directory\Read->validatePath()
#2 /home/cloudpanel/htdocs/www.performance-pcs.com/vendor/magento/framework/Code/GeneratedFiles.php(123): Magento\Framework\Filesystem\Directory\Read->isExist()
#3 /home/cloudpanel/htdocs/www.performance-pcs.com/vendor/magento/framework/Code/GeneratedFiles.php(98): Magento\Framework\Code\GeneratedFiles->isCleanGeneratedFilesAllowed()
#4 /home/cloudpanel/htdocs/www.performance-pcs.com/vendor/magento/framework/App/ObjectManagerFactory.php(114): Magento\Framework\Code\GeneratedFiles->cleanGeneratedFiles()
#5 /home/ in /home/cloudpanel/htdocs/www.performance-pcs.com/vendor/magento/framework/App/ObjectManager.php on line 36

2. Screenshots
image
image

### Things we tried and results:

1. Our Host of AWS Cloud server adjusted these settings in this order:
     a. From the error shown you have reached the limit of 768M that was set on the php config file. 
         We have increased it and set to memory_limit = 1024M 
     b. The php-fpm of cli had memory_limit = 4048M. We have set this to unilimited 
     c. So we have increased the memory limit to :8092M on file ".user.ini" on Magento root & inside pub/ location.
         mgt@performance-pcs:/home/cloudpanel/htdocs/www.performance-pcs.com/pub$ cat .user.ini
         memory_limit = 8092M 
2. How we fixed: Well we didn't fix actually but we proved to ourselves that it was the number of images that caused the problem. So we moved all our images out of media except for about 200 or so. And vola! The media gallery opened fine with the proper tree structure, search and sort fields and all functions worked.

image

So now what?? We need to access all our **NON-PRODUCT IMAGES!**. I can turn back on old media gallery and it does work but I would like to use the new one with the Adobe image features, that is handy!
---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [X ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### Additional info from Engcom
To reproduce the issue 
1. Execute this file to generate empty txt files in media folder https://gist.github.com/Nazar65/61c756ce76afa8c6718bf670794d548c execute in CMD bash glob.sh
2. Set memory_limit in php to 100MB
3. Open Media Galley

Actual Result: Media gallery directory tree component renders directory tree component

### PR DESCRIPTION
**Pull request by nazarklovanych**
REOPENED OLD PR -> https://github.com/magento/magento2/pull/30629

### Description (*)

Scan only directories for directory tree component
The result of using glob instead of iterator with more than 500000 files in media folder

2020-10-23-225235_1920x1080_scrot

### Fixed Issues (if relevant)

1. Fixes magento/magento2#30596

### Manual testing scenarios (*)
1. A lot of images in filesystem -> execute this file to generate empty txt files in media folder https://gist.github.com/Nazar65/61c756ce76afa8c6718bf670794d548c execute in CMD bash glob.sh
2. Set memory_limit in php to 100MB
2. Open Media Galley
3. Media gallery directory tree component  renders directory tree component

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by nazarklovanych** [comment]
@magento run all tests

**Comment by nazarklovanych** [comment]
https://github.com/magento run all tests

**Comment by nazarklovanych** [comment]
@magento run all tests

**Comment by nazarklovanych** [comment]
@magento run all tests

**Comment by nazarklovanych** [comment]
@magento run all tests

**Comment by nazarklovanych** [comment]
@magento run all tests

**Comment by nazarklovanych** [comment]
@magento run Integration Tests

**Comment by nazarklovanych** [comment]
@magento run all tests

**Comment by nazarklovanych** [comment]
@magento run all tests

**Comment by nazarklovanych** [comment]
@magento run all tests

**Comment by nazarklovanych** [comment]
@magento run all tests

**Comment by nazarklovanych** [comment]
Hello there :)  There is two tests failing about AWS s3 storage, but i don't have an S3 remote storage to recheck what is wrong there, and looks like all other failed tests is not related to my changes

**Comment by perfpcs** [comment]
Hello
After all these various fixes by different people, I still have no directory tree even with tour files listed here. I was the one that made the original PR and even though everything else now loads in admin like filters, search etc, the directory tree still does not load for me. I have like 2.7M images. I have tested by moving like over half the images out of the media folder, the tree will generate but not with 2.7M images. I am unsure the exact number of images where the tree but after over a year this is still not fixed and is VERY ANNOYING!

I see that Commerce edition only loads the WYSIWYG and catalog directories in media gallery. How can we set this in commerce edition to see if that fixes issue? I have so many .thumbs and original files in folders where an image optimizer extension creates WEBP images for site but dumps the originals in folders. And there are images for phone, tablet and desktop. That is why I have so many 2.7M.

I need some sort of fix to get the tree back. Suggestions??

**Comment by kimurabanyan** [comment]
Hello,

I tried this code and found the problem.
First of all, the pattern used in "isPathExcluded->execute" would be as follows if it is Magento default.

```
 /^(wysiwyg\b(?!-)(?:\/?[^\/]+)*\/?$|catalog[\/]+category\b(?!-)(?:\/?[^\/]+)*\/?$)/.
```

If $relativePath ='catalog', this pattern does not match, so getDirectories is recursively called in the following if statement.

https://github.com/magento/magento2/pull/35404/files#diff-f9876dbb5d6835b068311c3353b9ea2d0b535c7069bf474429822a3db8c960e9R81

The problem here is that under "catalog" there is not only a "category" but also a "product" directory.
In the recursively called getDirectories, $relativePath = 'catalog/category' will not be in the same if statement, but 'catalog/product' will be in the if statement and getDirectories will be called again.
In other words, getDirectories will be executed in all directories under 'catalog/product'.

This has a huge impact, and the more products we have, the more impact we will have.

**Comment by perfpcs** [comment]
@kimurabanyan were your comments directed to my issues with the code or @nazarklovanych issues with AWS S3 storage?

And did you propose a fix for this issue? Sorry was not clear to me.
thanks

**Comment by interwijs** [comment]
This is also a problem in Magento 2.4.4. Directory tree is missing. When I temporarily move folder pub/media/catalog/product the directory tree is loading just fine when old media gallery is disabled in the configuration. But this is no solution of course. The mentioned patches don't work either. Is there an other way to make sure the pub/media/catalog/product directories are excluded in Magento 2.4.4 when the directory tree is buildup?

**Comment by kimurabanyan** [comment]
@perfpcs 
 I commented because I found a performance issue in this PR code. It seems to be related to the problem you have.
 I haven't suggested a fix for this issue. I'm not currently using Adobe Stock, so I'm using the old media gallery to work around this issue.

**Comment by perfpcs** [comment]
@kimurabanyan  OK I understand you. Thanks for clarifying. I guess Magento has no one on their team that can fix this properly because most everyone I know all reverted back to old media gallery because the new one is non-functional. Sad!

**Comment by interwijs** [comment]
@perfpcs, when using the old media gallery I also experience problems in Magento 2.4.4. Images are not visible and the directory tree only shows Storage Root directory with 2 empty subdirectories with no name. When I try to upload an image I get the message: We can't upload the file to the current folder right now. Please try another folder. And when I try to add a directory I get the message: We cannot create the folder under the selected directory. So the old media gallery is also useless right now.

**Comment by perfpcs** [comment]
@interwijs I have not yet reverted to old media gallery in 2.4.4 but I will try it and see if it also is broke. Seems hard to believe if this is so that there is not more uproar from the community with no workable media gallery. And after 1.5 years of them knowing the issues, still no viable fix. Yet they are worried about merging code that does not even fix the problem. All this is hard to understand.

**Comment by pmonosolo** [comment]
> Hello,
> 
> I tried this code and found the problem. First of all, the pattern used in "isPathExcluded->execute" would be as follows if it is Magento default.
> 
> ```
>  /^(wysiwyg\b(?!-)(?:\/?[^\/]+)*\/?$|catalog[\/]+category\b(?!-)(?:\/?[^\/]+)*\/?$)/
> ```
> 
> If $relativePath ='catalog', this pattern does not match, so getDirectories is recursively called in the following if statement.
> 
> https://github.com/magento/magento2/pull/35404/files#diff-f9876dbb5d6835b068311c3353b9ea2d0b535c7069bf474429822a3db8c960e9R81
> 
> The problem here is that under "catalog" there is not only a "category" but also a "product" directory. In the recursively called getDirectories, $relativePath = 'catalog/category' will not be in the same if statement, but 'catalog/product' will be in the if statement and getDirectories will be called again. In other words, getDirectories will be executed in all directories under 'catalog/product'.
> 
> This has a huge impact, and the more products we have, the more impact we will have.

UGH...it looks like this is the culprit - this folder needs to be blocked off from being searched as there's no need to access it - its completely automated by the product creation / import processes.

**Comment by sprankhub** [comment]
This PR fixed the issues for us. Thanks, @nazarklovanych!

**Comment by perfpcs** [comment]
Still broken for me on 2.4.5-p1

https://github.com/magento/magento2/issues/30596

**Comment by sprankhub** [comment]
@perfpcs, yes, it is broken in 2.4.5-p1. However, did you try to apply the changes from this PR manually? Did you check if the changes from this PR solve the issue for you?

**Comment by perfpcs** [comment]
@sprankhub , yes I tried and am using this PR but it only partially works. Editing directory.xml results in no changes to the tree whatsoever so we can never get ONLY the WYSIWYG directory. Also we are missing many of the things that make this useful like the search box, sorting drop downs, etc.

image

Do you have all these things showing?

**Comment by sprankhub** [comment]
> yes I tried and am using this PR but it only partially works.

Please mind that this PR is about the performance / memory issue with large media directories. If you have different issues, you might want to look into other / open new issues.

> Editing directory.xml results in no changes to the tree whatsoever so we can never get ONLY the WYSIWYG directory.

This is not part of this PR.

> Also we are missing many of the things that make this useful like the search box, sorting drop downs, etc.

I do see all this:

image

**Comment by engcom-Hotel** [comment]
Hello @nazarklovanych,

Have you got a chance to resolve the conflicts of this PR?

Thanks

**Comment by engcom-Hotel** [comment]
Hi @nazarklovanych,

Thank you for your contribution!

As per the comment here https://github.com/magento/magento2/pull/35404#issuecomment-1614377153 we are unable to move further on this PR due to conflicts. Can you please look into it? 

Till then we are closing this PR.

Please feel free to reopen or let us know when you are ready to work on it again, and we will be happy to re-open it.

Thank you!
